packages/react-router/lib/router/router.ts

```diff
@@ -4940,32 +4940,46 @@ const lazyRouteFunctionCache = new WeakMap<
  * shouldRevalidate) and update the routeManifest in place which shares objects
  * with dataRoutes so those get updated as well.
  */
-async function loadLazyRoute(
+function loadLazyRoute(
   route: AgnosticDataRouteObject,
+  type: "loader" | "action",
   manifest: RouteManifest,
   mapRouteProperties: MapRoutePropertiesFunction
-) {
+): {
+  lazyRoutePromise: Promise<void> | undefined;
+  lazyHandlerPromise: Promise<void> | undefined;
+} {
   let routeToUpdate = manifest[route.id];
   invariant(routeToUpdate, "No route found in manifest");
 
   if (!route.lazy) {
-    return;
+    return {
+      lazyRoutePromise: undefined,
+      lazyHandlerPromise: undefined,
+    };
   }
 
   if (typeof route.lazy === "function") {
     // Check if we have a cached promise from a previous call
     let cachedPromise = lazyRouteFunctionCache.get(routeToUpdate);
     if (cachedPromise) {
-      await cachedPromise;
-      return;
+      return {
+        lazyRoutePromise: cachedPromise,
+        lazyHandlerPromise: cachedPromise,
+      };
     }
 
     // We use `.then` to chain additional logic to the lazy route promise so that
     // the consumer's lazy route logic is coupled to our logic for updating the
     // route in place in a single task. This ensures that the cached promise
     // contains all logic for managing the lazy route. This chained promise is
     // then awaited so that consumers of this function see the updated route.
-    let lazyRoutePromise = route.lazy().then((lazyRoute) => {
+    let lazyRoutePromise = (async () => {
```

Comment:
Since `loadLazyRoute` is no longer an async function, and is instead a sync function that returns an object with multiple promises, we need to wrap the `route.lazy()` call within an async IIFE so that errors are handled correctly.